routing/payment_lifecycle_test.go

```diff
@@ -491,3 +662,943 @@ func TestDecideNextStep(t *testing.T) {
+//
+//nolint:paralleltest
+func TestResumePaymentRetry(t *testing.T) {
+	// Create a test paymentLifecycle with the initial two calls mocked.
+	p, m := setupTestPaymentLifecycle(t)
+
+	// Create a dummy route that will be returned by `RequestRoute`.
+	paymentAmt := lnwire.MilliSatoshi(10000)
+	rt := createDummyRoute(t, paymentAmt)
+
+	// We now enter the payment lifecycle loop.
+	//
+	// 1.1. calls `FetchPayment` and return the payment.
+	m.control.On("FetchPayment", p.identifier).Return(m.payment, nil).Once()
+
+	// 1.2. calls `GetState` and return the state.
+	ps := &channeldb.MPPaymentState{
+		RemainingAmt: paymentAmt,
+	}
+	m.payment.On("GetState").Return(ps).Once()
+
+	// NOTE: GetStatus is only used to populate the logs which is
+	// not critical so we loosen the checks on how many times it's
+	// been called.
+	m.payment.On("GetStatus").Return(channeldb.StatusInFlight)
+
+	// 1.3. decideNextStep now returns stepProceed.
+	m.payment.On("AllowMoreAttempts").Return(true, nil).Once()
+
+	// 1.4. mock requestRoute to return an route.
+	m.paySession.On("RequestRoute",
+		ps.RemainingAmt, p.feeLimit, uint32(ps.NumAttemptsInFlight),
+		uint32(p.currentHeight),
+	).Return(rt, nil).Once()
+
+	// 1.5. mock `registerAttempt` to return an attempt.
+	//
+	// Mock NextPaymentID to always return the attemptID.
+	attemptID := uint64(0)
+	p.router.cfg.NextPaymentID = func() (uint64, error) {
+		return attemptID, nil
+	}
+
+	// Mock shardTracker to return the mock shard.
+	m.shardTracker.On("NewShard",
+		attemptID, false,
+	).Return(m.shard, nil).Run(func(args mock.Arguments) {
+		// Once this is called, increment the attemptID for the next
+		// call.
+		attemptID++
+	}).Once()
+
+	// Mock the methods on the shard.
+	m.shard.On("MPP").Return(&record.MPP{}).Twice().
+		On("AMP").Return(nil).Once().
+		On("Hash").Return(p.identifier).Once()
+
+	// Mock the time and expect it to be called.
+	m.clock.On("Now").Return(time.Now())
+
+	// We now register attempt and return no error.
+	m.control.On("RegisterAttempt",
+		p.identifier, mock.Anything,
+	).Return(nil).Run(func(args mock.Arguments) {
+		ps.RemainingAmt = paymentAmt / 2
+	}).Once()
+
+	// 1.6. mock `sendPaymentAttempt` to succeed, which brings us into the
+	// next iteration of the lifecycle where we mock a temp failure.
+	m.payer.On("SendHTLC",
+		mock.Anything, attemptID, mock.Anything,
+	).Return(nil).Once()
+
+	// We now enter the second iteration of the lifecycle loop.
+	//
+	// 2.1. calls `FetchPayment` and return the payment.
+	m.control.On("FetchPayment", p.identifier).Return(m.payment, nil).Once()
+
+	// 2.2. calls `GetState` and return the state.
+	m.payment.On("GetState").Return(ps).Once()
+
+	// 2.3. decideNextStep now returns stepProceed so we can send the
+	// second attempt.
+	m.payment.On("AllowMoreAttempts").Return(true, nil).Once()
+
+	// 2.4. mock requestRoute to return an route.
+	m.paySession.On("RequestRoute",
+		paymentAmt/2, p.feeLimit, uint32(ps.NumAttemptsInFlight),
```

Comment:
because `paymentAmt` is the control variable used in the test, and `ps.RemainingAmt` is a state under testing. For instance, maybe the first attempt sends more than half, if we use `ps.RemainingAmt` then it won't get caught, as the mockers will assert the specified variables are called as expected.

Reply:
actually i was wrong, `ps` is also a controlling variable here so switched